### PR TITLE
Allow post and basic client authentication mode for OIDC

### PR DESCRIPTION
**Pull request by bhugot**
Introduce the possibility to configure the client authentication mode on access_token endpoint https://tools.ietf.org/html/rfc6749#section-2.3.1

this is another proposal that not break the current behavior and allow basic auth for 
Encode ClientId/Secret in Basic Authorization Header per OIDC Spec #1738

### CODE REVIEW

**Comment by Tratcher** [comment]
Tests?

What servers require this that don't work with the current version?

File: src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs [hunk 1]
```diff
@@ -738,6 +738,14 @@ private void PopulateSessionProperties(OpenIdConnectMessage message, Authenticat
             Logger.RedeemingCodeForTokens();
 
             var requestMessage = new HttpRequestMessage(HttpMethod.Post, _configuration.TokenEndpoint);
+
+            if (this.Options.ClientAuthenticationMode == OpenIdConnectClientAuthenticationMode.Basic)
+            {
+                requestMessage.Headers.Authorization = new AuthenticationHeaderValue("Basic", Convert.ToBase64String(Encoding.UTF8.GetBytes(tokenEndpointRequest.ClientId+":"+tokenEndpointRequest.ClientSecret)));
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs]: Separate into multiple statements.

**Comment by bhugot** [comment]
this implementation by default need it
https://www.forgerock.com/ , I tested it on internal instance.

**Comment by Tratcher** [comment]
A feature like this needs unit tests. See https://github.com/aspnet/Security/tree/dev/test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect

**Comment by Tratcher** [comment]
@blowdart for feature triage.

**Comment by blowdart** [comment]
We'd need to go beyond unit tests to functional tests against a compliant OIDC service which implements basic auth.

However we're not adding major features like this until v3, so I'm putting it in the v3 milestone for now.

**Comment by Tratcher** [comment]
Note that functional testing would be manual on our part, that's how we do it for all of the other providers.

File: src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs [hunk 2]
```diff
@@ -738,6 +738,16 @@ private void PopulateSessionProperties(OpenIdConnectMessage message, Authenticat
             Logger.RedeemingCodeForTokens();
 
             var requestMessage = new HttpRequestMessage(HttpMethod.Post, _configuration.TokenEndpoint);
+
+            if (this.Options.ClientAuthenticationMode == OpenIdConnectClientAuthenticationMode.Basic)
+            {
+                var basicHeader = Convert.ToBase64String(
+                    Encoding.UTF8.GetBytes($"{tokenEndpointRequest.ClientId}:{tokenEndpointRequest.ClientSecret}"));
```
[the remark this answers is not in the document]
Comment (bhugot) [reply src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs]: It s the way it is with openid. So why not having it and make it work for everyone

[the remark this answers is not in the document]
Comment (PinpointTownes) [reply src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs]: That’s my point: your PR doesn’t match what the specification says.
To make it compliant, you’ll need to use formURL-encoding.

**Comment by bhugot** [comment]
Look better it s what it s doing by default

**Comment by PinpointTownes** [comment]
> Look better it s what it s doing by default

If you read the specification, you'll see that's not what your PR is doing:

> **The client identifier is encoded using the "application/x-www-form-urlencoded" encoding algorithm per Appendix B**, and the encoded value is used as the username; the client password is encoded using the same algorithm and used as the password.

You MUST encode both the username and the password using the form-urlencoded format before sending them to the server.

(I agree the wording is a bit confusing... and that's why most server or client implementations have been historically non-standard-compliant).

**Comment by Eilon** [comment]
Triage discussion:

Hi, we're closing this PR because next week we're moving all the code to a different repo and there's no way to move over PRs. But, we do like this feature and would like to work with you to get it in after we move the code. If that sounds good to you, please let us know, and we will work with you on next steps (such as a few code cleanup things).

**Comment by bhugot** [comment]
Hi, feel free to contact me when your done

**Comment by samvik** [comment]
Hello, any idea of when this pr can be merged? Let me know if there are anything I can do to help? Am in the process of integrate against an openid provider which only supports basic auth.

**Comment by Tratcher** [comment]
@bhugot the code has moved to https://github.com/aspnet/aspnetcore. The experience is still a bit rough, but it would be possible to cherry pick your PR over there.

**Comment by bhugot** [comment]
@Tratcher did the PR aspnet/aspnetcore#4533

**Pull request closed without merging**